mod_amity/amity.py

```diff
@@ -0,0 +1,218 @@
+                if len(self.living_spaces["available"]) > 0 else None
+            if living_space is not None and person.accommodation == 'Y':
+                living_space.allocate_space(person)
+                person.assign_living_space(living_space.name)
+
+        self.check_person_allocation(person)
+        self.check_room_availability()
+
+        return person
+
+    def get_unallocated_persons(self):
+        """
+        gets persons not fully allocated spaces
+        :return: dict with  fellows and staff
+        """
+        return dict(staff=list(set(self.staff) - set(self.allocated_staff)),
+                    fellows=list(set(self.fellows) - set(self.allocated_fellows)))
+
+    def get_rooms(self, room_name=None):
+        """
+        gets all offices and living spaces in the allocation pool
+        :param room_name: (optional) filter the result by name
+        :return: dict of all living_spaces and offices
+        """
+        offices = self.offices["unavailable"] + self.offices["available"]
+        living_spaces = self.living_spaces["unavailable"] + self.living_spaces["available"]
+        rooms = offices + living_spaces
+        if room_name is not None:
+            for room in rooms:
+                if room.name == room_name:
+                    return room
+        else:
+            return dict(living_spaces=living_spaces, offices=offices)
+
+    def get_staff_id(self):
+        """
+        generate unique ids for staff
+        :return: staff id i.e ST001
+        """
+        staff_id = self.ids["staff"][0] + 1
+        self.ids["staff"][0] = staff_id
+        return "ST{0}".format(str(staff_id).rjust(3, '0'))
+
+    def generate_fellow_id(self):
+        """
+        generate unique ids for fellow
+        :return: staff id i.e FL001
+        """
+        staff_id = self.ids["fellow"][0] + 1
+        self.ids["fellow"][0] = staff_id
+        return "FL{0}".format(str(staff_id).rjust(3, '0'))
+
+    def check_person_allocation(self, person):
+        """
+        checks if fellow/staff is fully allocated and add to allocated list
+        :param person: an instance of fellow or staff
+        :return:
+        """
+        if person.role == Role.STAFF:
+            if person.office is not None:
+                self.allocated_staff.append(person)
+
+        elif person.role == Role.FELLOW:
+            if person.living_space is not None and person.office is not None:
+                self.allocated_fellows.append(person)
+
+    def find_person_by_name(self, name):
+        """
+        Search combined list of fellows and staff for partial match of name
+        :param name:
+        :return: List of person object matching name
+        """
+        match = []
+        for person in (self.fellows + self.staff):
+            if name in person.name:
+                match.append(person)
+        return match
+
+    def find_person_by_id(self, person_id):
+        """
```

Comment:
ideally the function returns the person instance when the id is found, to be used in the docopts for display. however, great idea on using oneliners